Add ability to select to install matching version already available on the current image

nikolai-laevskii opened the pull request:
**Description:**
Adds ability to set preference for already installed versions when setting up dotnet environment. If `prefer-installed` flag is active, action will attempt to check already installed versions with dotnet utility. If matching version exists on the runner, request for this specific version will be passed to the install-script, which has ability to verify already installed version and skip download if current installation is not broken.

**Related issue:**
https://github.com/actions/setup-dotnet/issues/356

**Check list:**
- [x] Mark if documentation changes are required.
- [x] Mark if tests were added or updated to cover the changes.

nikolai-laevskii committed:
Add test action for listing sdks

nikolai-laevskii committed:
Update install scripts

nikolai-laevskii committed:
Add feature of prefering already installed versions

nikolai-laevskii committed:
Update integration tests

nikolai-laevskii committed:
Update e2e tests

nikolai-laevskii committed:
Update e2e tests

nikolai-laevskii committed:
Add additional e2e test

nikolai-laevskii committed:
Update listSdks to not fail when dotnet util is not installed

nikolai-laevskii committed:
Update build

nikolai-laevskii committed:
Update dotnet utils, to handle more cases

nikolai-laevskii committed:
Add unit tests

nikolai-laevskii committed:
Move e2e test to appropriate workflow

nikolai-laevskii committed:
Fix typo in e2e test

nikolai-laevskii committed:
Eliminate hard dependency on version resolver in installer

nikolai-laevskii committed:
Format & build

nikolai-laevskii committed:
Update tests according to new api

dsame commented on src/dotnet-utils.ts:
```diff
@@ -0,0 +1,49 @@
+ * correct version number like '3.1.201', '3.1.201', '3.1.201', '3.1.201', '6.0.402'
+ */
+export const matchVersionToList = (version: string, versions: string[]) => {
+  if (!version || version === 'x' || version === '*') {
```
Is not `version` argument is semver specification and semver matcher [package](https://www.npmjs.com/package/semver) can be used in order to match? Also `version` can be renamed to versionSpec then.

nikolai-laevskii replied:
This is a good point, but semver doesn't seem to work correctly with `6.0.4xx` version format, i.e:
```typescript
semver.satisfies('6.0.402', '6.0.4xx') // -> false (should be true)
```

dsame commented on src/dotnet-utils.ts:
```diff
@@ -0,0 +1,49 @@
+ * '3.1', '3.1.x', '3', '3.x', '6.0.4xx' to
+ * correct version number like '3.1.201', '3.1.201', '3.1.201', '3.1.201', '6.0.402'
+ */
+export const matchVersionToList = (version: string, versions: string[]) => {
```
I would either documented return with [typdoc](https://typedoc.org/tags/returns/) notation or use typescript typing.

Currently it is not clear what the function is expected to return: array, element, boolean, index or undefined.

nikolai-laevskii replied:
I'll set return type explicitly, fair point

dsame commented on src/installer.ts:
```diff
@@ -33,11 +34,30 @@ export class DotnetVersionResolver {
+      installedVersions
+    );
+
+    if (matchingInstalledVersion === undefined) {
```
May be it is better to replace `if (matchingInstalledVersion === undefined)`  with `if (matchingInstalledVersion)` and exchange the branches?

dsame commented on src/installer.ts:
```diff
@@ -59,9 +79,9 @@ export class DotnetVersionResolver {
+  private createVersionArgument(updatedVersion?: string) {
     this.resolvedArgument.type = 'version';
-    this.resolvedArgument.value = this.inputVersion;
+    this.resolvedArgument.value = updatedVersion ?? this.inputVersion;
```
I'd forced the argument to be defined (remove "?" mark). The updatedVersion is already verified outside and can be guarded by compile-time check. This avoid unnecessary run-time check.

nikolai-laevskii replied:
Fair

dsame commented on src/installer.ts:
```diff
@@ -253,12 +273,12 @@ export class DotnetCoreInstaller {
 
-  constructor(private version: string, private quality: QualityOptions) {}
+  constructor(
+    private dotnetVersion: DotnetVersion,
```
This can be made read only

nikolai-laevskii committed:
Address review comments
- add explicit return type for findMatching version function
- force version argument to be defined in createVersionArgument
- make dotnetVersion and quality properties readonly

nikolai-laevskii committed:
Update installers
